<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Retrieval-Augmented-Generative-Question-Answering-(GQA)" data-toc-modified-id="Retrieval-Augmented-Generative-Question-Answering-(GQA)-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Retrieval-Augmented Generative Question Answering (GQA)</a></span><ul class="toc-item"><li><span><a href="#Libaries" data-toc-modified-id="Libaries-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Libaries</a></span></li><li><span><a href="#Set-API" data-toc-modified-id="Set-API-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Set API</a></span><ul class="toc-item"><li><span><a href="#Exkurs:-Ask-about-HSLU-without-knowledgebase" data-toc-modified-id="Exkurs:-Ask-about-HSLU-without-knowledgebase-1.2.1"><span class="toc-item-num">1.2.1&nbsp;&nbsp;</span>Exkurs: Ask about HSLU without knowledgebase</a></span></li></ul></li><li><span><a href="#Load-data" data-toc-modified-id="Load-data-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Load data</a></span></li><li><span><a href="#Building-a-Knowledge-Base:-Embeed-data" data-toc-modified-id="Building-a-Knowledge-Base:-Embeed-data-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Building a Knowledge Base: Embeed data</a></span></li><li><span><a href="#Create-Chunks" data-toc-modified-id="Create-Chunks-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Create Chunks</a></span></li><li><span><a href="#Create-Embeedings-(ada-002)" data-toc-modified-id="Create-Embeedings-(ada-002)-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Create Embeedings (ada-002)</a></span></li><li><span><a href="#Vector-database" data-toc-modified-id="Vector-database-1.7"><span class="toc-item-num">1.7&nbsp;&nbsp;</span>Vector database</a></span></li><li><span><a href="#Indexing" data-toc-modified-id="Indexing-1.8"><span class="toc-item-num">1.8&nbsp;&nbsp;</span>Indexing</a></span></li><li><span><a href="#Query-embeeding-and-get-context" data-toc-modified-id="Query-embeeding-and-get-context-1.9"><span class="toc-item-num">1.9&nbsp;&nbsp;</span>Query embeeding and get context</a></span></li><li><span><a href="#Set-prompt-and-results" data-toc-modified-id="Set-prompt-and-results-1.10"><span class="toc-item-num">1.10&nbsp;&nbsp;</span>Set prompt and results</a></span><ul class="toc-item"><li><span><a href="#Example-1:-German" data-toc-modified-id="Example-1:-German-1.10.1"><span class="toc-item-num">1.10.1&nbsp;&nbsp;</span>Example 1: German</a></span></li><li><span><a href="#Example-2:-French" data-toc-modified-id="Example-2:-French-1.10.2"><span class="toc-item-num">1.10.2&nbsp;&nbsp;</span>Example 2: French</a></span></li><li><span><a href="#Example-3:-German" data-toc-modified-id="Example-3:-German-1.10.3"><span class="toc-item-num">1.10.3&nbsp;&nbsp;</span>Example 3: German</a></span></li><li><span><a href="#Example-4:-Chinese" data-toc-modified-id="Example-4:-Chinese-1.10.4"><span class="toc-item-num">1.10.4&nbsp;&nbsp;</span>Example 4: Chinese</a></span></li><li><span><a href="#Example-5:-German-again" data-toc-modified-id="Example-5:-German-again-1.10.5"><span class="toc-item-num">1.10.5&nbsp;&nbsp;</span>Example 5: German again</a></span></li><li><span><a href="#Example-6:-Spanisch" data-toc-modified-id="Example-6:-Spanisch-1.10.6"><span class="toc-item-num">1.10.6&nbsp;&nbsp;</span>Example 6: Spanisch</a></span></li><li><span><a href="#Example7:-Japanese" data-toc-modified-id="Example7:-Japanese-1.10.7"><span class="toc-item-num">1.10.7&nbsp;&nbsp;</span>Example7: Japanese</a></span></li><li><span><a href="#Example-8:-Swedish" data-toc-modified-id="Example-8:-Swedish-1.10.8"><span class="toc-item-num">1.10.8&nbsp;&nbsp;</span>Example 8: Swedish</a></span></li><li><span><a href="#Outcome-withouth-primer-for-correct-answers-(Example-1)" data-toc-modified-id="Outcome-withouth-primer-for-correct-answers-(Example-1)-1.10.9"><span class="toc-item-num">1.10.9&nbsp;&nbsp;</span>Outcome withouth primer for correct answers (Example 1)</a></span></li></ul></li></ul></li><li><span><a href="#Gradio:-HSLU-multilingual-project-Chatbot-(retrieval-augmentet-GQA)" data-toc-modified-id="Gradio:-HSLU-multilingual-project-Chatbot-(retrieval-augmentet-GQA)-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Gradio: HSLU multilingual project Chatbot (retrieval-augmentet GQA)</a></span></li></ul></div>

# Retrieval-Augmented Generative Question Answering (GQA)

## Libaries

In [2]:
import pandas as pd
import openai
from langchain.embeddings.openai import OpenAIEmbeddings
import tiktoken
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain.text_splitter import CharacterTextSplitter
import pinecone
from tqdm.auto import tqdm
from uuid import uuid4
from IPython.display import Markdown
import gradio as gr
import warnings
warnings.filterwarnings("ignore")

## Set API

In [3]:
# Set API key
openai.api_key = os.environ["OPENAI_API_KEY"] = getpass("Paste your OpenAI API key here and hit enter:")

### Exkurs: Ask about HSLU without knowledgebase

In [42]:
query = "Welche Projekte gibt es auf der HSLU in Luzern zum Thema Senioren?"
res = openai.Completion.create(
    engine='text-davinci-003',
    prompt=query,
    temperature=0,
    max_tokens=400,
    top_p=1,
    frequency_penalty=0,
    presence_penalty=0,
    stop=None)

res['choices'][0]['text'].strip()

'Auf der HSLU in Luzern gibt es verschiedene Projekte, die sich mit dem Thema Senioren beschäftigen. Ein Beispiel ist das Projekt „Aktiv im Alter“, das sich mit der Förderung der Lebensqualität älterer Menschen befasst. Es bietet verschiedene Aktivitäten an, wie z.B. Kurse, Workshops und Vorträge. Ein weiteres Projekt ist das „Senioren-Netzwerk“, das älteren Menschen eine Plattform bietet, um sich auszutauschen und zu vernetzen. Es organisiert regelmässig Veranstaltungen, wie z.B. Vorträge, Seminare und Workshops. Darüber hinaus gibt es auch das Projekt „Senioren-Beratung“, das älteren Menschen bei der Bewältigung ihrer Alltagsprobleme hilft. Es bietet Beratung und Unterstützung in verschiedenen Bereichen, wie z.B. Finanzen, Gesundheit und soziale Kontakte.'

## Load data

In [4]:
# Load data
df1 = pd.read_excel("4_preprocess.xlsx")
# Create dataframe
df1 = pd.DataFrame(df1, columns=['Projekt ID', "Projekttitel", "Teaser",  "Abstract", "url_new", "Projektstatus", "year_start", "year_end", "SAP Interne Organisationen", "corpus"])

In [5]:
# Rename text column
df1.rename(columns={"corpus": "text"}, inplace=True)

In [6]:
# Convert dataframe to a list of dict for Pinecone data upsert
data = df1.to_dict('records')

In [7]:
# Inspect
data[6]

{'Projekt ID': 78839,
 'Projekttitel': 'Geschwister von Menschen mit einer Behinderung oder einer chronischen Erkrankung',
 'Teaser': 'Die Lebenssituation von Geschwistern von Menschen mit einer Behinderung in der Schweiz wird untersucht mit dem Ziel ihnen die notwendige Unterstützung zu ermöglichen. ',
 'Abstract': '    Aufgrund von Hinweisen aus der Forschung kann angenommen werden, dass sich Geschwister von Menschen mit einer Behinderung oder einer chronischen Krankheit in einer Lebenssituation mit tieferer Lebensqualität und einem erhöhten Risiko für Entwicklungsbeeinträchtigung befinden. Die Hochschule Luzern (HSLU) untersucht im Auftrag des Vereins Raum für Geschwister VRG und der Stiftung KiFa Schweiz die Lebenssituation von Geschwistern in der Schweiz. Das Ziel ist es, die gesellschaftliche Relevanz sichtbar zu machen. Ergebnisse der Untersuchung sollen dazu beitragen, dass Geschwister von Kindern und Erwachsenen mit Beeinträchtigung in einer entwicklungsfördernden Lebenssituat

In [8]:
new_data = data

In [9]:
new_data[2]

{'Projekt ID': 54075,
 'Projekttitel': 'Digitalisierung und Effizienzverbesserung bei Hypothekarprozessen',
 'Teaser': 'Banken sehen sich gezwungen, nicht nur an der Kundenschnittstelle Anpassungen vorzunehmen, sondern vor allem auch die Prozesse effizienter und digitaler zu machen. ',
 'Abstract': '        Im Rahmen dieser Untersuchung haben wir analysiert, wie sich der traditionelle Hypothekarprozess heute präsentiert und wo und in welcher Höhe Potenziale für Kosteneinsparungen liegen. Die Analyse dieser Studie beschränkt sich auf den Hypothekarprozess, da dieser vergleichsweise lange dauert und häufig durchlaufen wird. In Zusammenarbeit mit dem Digitalisierungs- und Security-Produkte Anbieter ti&m wurden die Hypothekarprozesse von fünf mittelgrossen Schweizer Retail Banken mittels ausführlichen Interviews unter die Lupe genommen. Die Analyse der Kreditprozesse beschränkt sich auf Hypotheken für Privatpersonen und den Standardfall, dass der Kunde für den Neuabschluss den persönlichen

## Building a Knowledge Base: Embeed data

Example

In [10]:
embed_model = "text-embedding-ada-002"

res = openai.Embedding.create(
    input=[
        "Sample document text goes here",
        "there will be several phrases in each batch"
    ], engine=embed_model
)

In [11]:
# vector embeddings are stored within the 'data' key
res.keys()

dict_keys(['object', 'data', 'model', 'usage'])

In [12]:
# we have created two vectors (one for each sentence input)
len(res['data'])

2

In [13]:
# we have created two 1536-dimensional vectors
len(res['data'][0]['embedding']), len(res['data'][1]['embedding'])

(1536, 1536)

## Create Chunks

In [14]:
tokenizer = tiktoken.get_encoding('cl100k_base') 

# Create the length function
def tiktoken_len(text):
    tokens = tokenizer.encode(
        text,
        disallowed_special=()
    )
    return len(tokens)

In [15]:
text_splitter = RecursiveCharacterTextSplitter(
    chunk_size=400,
    chunk_overlap=20,
    length_function=tiktoken_len,
    separators=["\n\n", "\n", " ", ""]) 

In [16]:
chunks = text_splitter.split_text(data[6]['text'])[:3]
chunks

['Geschwister von Menschen mit einer Behinderung oder einer chronischen Erkrankung Die Lebenssituation von Geschwistern von Menschen mit einer Behinderung in der Schweiz wird untersucht mit dem Ziel ihnen die notwendige Unterstützung zu ermöglichen. Aufgrund von Hinweisen aus der Forschung kann angenommen werden, dass sich Geschwister von Menschen mit einer Behinderung oder einer chronischen Krankheit in einer Lebenssituation mit tieferer Lebensqualität und einem erhöhten Risiko für Entwicklungsbeeinträchtigung befinden. Die Hochschule Luzern (HSLU) untersucht im Auftrag des Vereins Raum für Geschwister VRG und der Stiftung KiFa Schweiz die Lebenssituation von Geschwistern in der Schweiz. Das Ziel ist es, die gesellschaftliche Relevanz sichtbar zu machen. Ergebnisse der Untersuchung sollen dazu beitragen, dass Geschwister von Kindern und Erwachsenen mit Beeinträchtigung in einer entwicklungsfördernden Lebenssituation aufwachsen und leben können und die dazu notwendige Unterstützung bek

In [18]:
tiktoken_len(chunks[0]), tiktoken_len(chunks[1])

(398, 256)

## Create Embeedings (ada-002)

In [19]:
import os
os.environ["OPENAI_API_KEY"] = "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"
#os.environ["COHERE_API_KEY"] = ""
#os.environ["HUGGINGFACEHUB_API_TOKEN"] = ""
#os.environ["SERPAPI_API_KEY"] = ""

In [20]:
from getpass import getpass
OPENAI_API_KEY = getpass("OpenAI API Key: ")

In [21]:
from langchain.embeddings.openai import OpenAIEmbeddings
model_name = 'text-embedding-ada-002' 
embed = OpenAIEmbeddings()

In [22]:
# embed
texts = [
    'this is the first chunk of text',
    'then another second chunk of text is here'
]
res = embed.embed_documents(texts)
len(res), len(res[0])
   

(2, 1536)

## Vector database

In [24]:
# Initialize connection to pinecone (get API key at app.pinecone.io)
pinecone.init(api_key="Pass your key", environment="us-west1-gcp")

index_name = 'lekgrag'

# If the index does not exist, we create it
if index_name not in pinecone.list_indexes():
    pinecone.create_index(
        index_name,
        dimension= len(res[0]), #dimension=shape[1], 
        metric='cosine')  ##dotproduct

In [25]:
# Connect to index
index = pinecone.Index(index_name)
index.describe_index_stats()

{'dimension': 1536,
 'index_fullness': 0.0,
 'namespaces': {},
 'total_vector_count': 0}

## Indexing

In [26]:
# Indexing
batch_limit = 100

texts = []
metadatas = []

for i, record in enumerate(tqdm(data)):
    # first get metadata fields for this record
    metadata = {
        'Projekt_ID': str(record['Projekt ID']),
        'source': record['url_new'],
        'title': record['Projekttitel'],
        'abstract': record['Abstract'],
    }
    # now we create chunks from the record text
    record_texts = text_splitter.split_text(record['text'])
    # create individual metadata dicts for each chunk
    record_metadatas = [{
        "chunk": j, "text": text, **metadata
    } for j, text in enumerate(record_texts)]
    # append these to current batches
    texts.extend(record_texts)
    metadatas.extend(record_metadatas)
    # if we have reached the batch_limit we can add texts
    if len(texts) >= batch_limit:
        ids = [str(uuid4()) for _ in range(len(texts))]
        embeds = embed.embed_documents(texts)
        index.upsert(vectors=zip(ids, embeds, metadatas))
        texts = []
        metadatas = []

  0%|          | 0/2735 [00:00<?, ?it/s]

In [67]:
# Inspect index
index.describe_index_stats()

{'dimension': 1536,
 'index_fullness': 0.0,
 'namespaces': {'': {'vector_count': 16599}},
 'total_vector_count': 16599}

## Query embeeding and get context


In [103]:
query = "welche projekte gibt es aktuell zu senioren?"

In [104]:
# Embeed Query
embed_model = "text-embedding-ada-002"

res = openai.Embedding.create(
    input=[query],
    engine=embed_model
)

# Retrieve from Pinecone
xq = res['data'][0]['embedding']

# Get relevant contexts (including the questions)
res = index.query(xq, top_k=3, include_metadata=True)

In [105]:
# Inspect
res

{'matches': [{'id': '8af35c80-78ee-4210-97d1-404d9c70011a',
              'metadata': {'Projekt_ID': '27721',
                           'abstract': 'Ein Team der Hochschule Luzern – '
                                       'Soziale Arbeit (HSLU - SA) arbeitet im '
                                       'Rahmen des Projektes „Integrierte '
                                       'Sozialplanung als Innovation für die '
                                       'Versorgung im Alter" (ISPInoVA) mit '
                                       'der Fachhochschule Köln zusammen. Mit '
                                       'dem Projekt verfolgt die HSLU – SA das '
                                       'Ziel, die Entwicklungen der sozialen '
                                       'Planung für ältere Menschen in '
                                       'Schweizer Städten und Gemeinden sowie '
                                       'deren Umsetzung zu erfassen und zu '
                               

In [106]:
# Retrieval function

# Set limit
limit = 3750

# Function
def retrieve(query):
    res = openai.Embedding.create(
        input=[query],
        engine=embed_model
    )

    # Retrieve from Pinecone
    xq = res['data'][0]['embedding']

    # Get relevant contexts
    res = index.query(xq, top_k=3, include_metadata=True)
    contexts = [
        x['metadata']['text'] for x in res['matches']
    ]

    # Build our prompt with the retrieved contexts included
    prompt_start = (
        "Answer the question based on the context below.\n\n"+
        "Context:\n"
    )
    prompt_end = (
        f"\n\nQuestion: {query}\nAnswer:"
    )
    # Append contexts until hitting limit
    for i in range(1, len(contexts)):
        if len("\n\n---\n\n".join(contexts[:i])) >= limit:
            prompt = (
                prompt_start +
                "\n\n---\n\n".join(contexts[:i-1]) +
                prompt_end
            )
            break
        elif i == len(contexts)-1:
            prompt = (
                prompt_start +
                "\n\n---\n\n".join(contexts) +
                prompt_end
            )
    return prompt

In [107]:
# Firstly retrieve relevant items from Pinecone
query_with_contexts = retrieve(query)
query_with_contexts

'Answer the question based on the context below.\n\nContext:\nan, die sich aktuell mit Themen der Planung für die ältere Bevölkerung beschäftigen.\n\n---\n\nentwickelt, welche sie nun weitgehend selbstorganisiert umsetzen. Die Ideen reichen von einem öffentlichen Bücherschrank, einem Erzähl- und verschiedenen Generationen-Cafés, einem Generationengarten und einem Generationenspielplatz bis zum Wochenmarkt und der Einführung von "netten Toiletten". Zur Finanzierung einiger grösserer Projekte wurde ein Finanzierungsgesuch bei der Age Stiftung eingereicht. Erwartete Ergebnisse Die Umsetzung der Projektideen soll die Lebensqualität für alle erhöhen, die Beteiligten bestärken und einen Beitrag zur Gestaltung des demographischen Wandels leisten.\n\n---\n\nsollen in einem vierten Teilprojekt die Bedürfnisse benachteiligter Gruppen, wie zum Beispiel älterer Migrantinnen und Migranten, ermittelt und in den Prozess miteinbezogen werden. Die Teilprojekte sollen eine Ver besserung der Lebensqualit

In [110]:
# Function to complete Prompt
def complete(prompt):
    # query text-davinci-003
    res = openai.Completion.create(
        engine='text-davinci-003',
        prompt=prompt,
        temperature=0,
        max_tokens=400,
        top_p=1,
        frequency_penalty=0,
        presence_penalty=0,
        stop=None
    )
    return res['choices'][0]['text'].strip()

In [111]:
# Complete the context-infused query
complete(query_with_contexts)

'Es gibt aktuell Projekte wie einen öffentlichen Bücherschrank, ein Erzähl- und Generationen-Café, einen Generationengarten, einen Generationenspielplatz, einen Wochenmarkt und die Einführung von "netten Toiletten".'

In [112]:
# Get list of retrieved text
contexts = [item['metadata']['text'] for item in res['matches']]
contexts_2 = [item['metadata']['source'] for item in res['matches']]
contexts_3 = [item['metadata']['title'] for item in res['matches']]
contexts_4 = [item['metadata']['abstract'] for item in res['matches']]

augmented_query = "\n\n".join(contexts)+"\n  \n".join(contexts_2)+"\n\n-----\n\n"+"\n  \n".join(contexts_3)+"\n\n-----\n\n"+"\n  \n".join(contexts_4)+"\n\n-----\n\n"+query

In [86]:
# Get list of retrieved text
#contexts = [item['metadata']['text'] for item in res['matches']]
#augmented_query = "\n\n---\n\n".join(contexts)+"\n\n-----\n\n"+query

In [83]:
print(augmented_query)    

an, die sich aktuell mit Themen der Planung für die ältere Bevölkerung beschäftigen.

entwickelt, welche sie nun weitgehend selbstorganisiert umsetzen. Die Ideen reichen von einem öffentlichen Bücherschrank, einem Erzähl- und verschiedenen Generationen-Cafés, einem Generationengarten und einem Generationenspielplatz bis zum Wochenmarkt und der Einführung von "netten Toiletten". Zur Finanzierung einiger grösserer Projekte wurde ein Finanzierungsgesuch bei der Age Stiftung eingereicht. Erwartete Ergebnisse Die Umsetzung der Projektideen soll die Lebensqualität für alle erhöhen, die Beteiligten bestärken und einen Beitrag zur Gestaltung des demographischen Wandels leisten.

sollen in einem vierten Teilprojekt die Bedürfnisse benachteiligter Gruppen, wie zum Beispiel älterer Migrantinnen und Migranten, ermittelt und in den Prozess miteinbezogen werden. Die Teilprojekte sollen eine Ver besserung der Lebensqualität älterer Menschen erzielen und aufzeigen, wie ihre Ressourcen besser gefördert

## Set prompt and results

### Example 1: German

In [115]:
# system message to 'prime' the model
primer = f"""You are a friendly Q&A bot. A highly intelligent system that answers user questions based only on the information provided by the user above each question. You can detect the language of the users questionn. You start your answer always in the following format according to the language detected: "At HSLU there are following projects related to {query}". Provide the identical title and summarizse the abstract from information provided by the user. Take note of the query language and provide anser in the same language as the querry. Take note of the sources and include them in the answer in the format: "SOURCES: source1 source2", use "SOURCES" in capital letters regardless of the number of sources. If the information can not be found in the information
provided by the user you truthfully say "I don't know. I use just the top 10 semantic search results, please have a look on the website: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/".
"""

res = openai.ChatCompletion.create( 
    model = "gpt-3.5-turbo", temperature=0,
    messages=[
        {"role": "system", "content": primer},
        {"role": "user", "content": augmented_query}])

In [116]:
from IPython.display import Markdown
display(Markdown(res['choices'][0]['message']['content']))

At HSLU there are following projects related to "welche projekte gibt es aktuell zu senioren?":

1. Project title: Altersplanung und Beteiligungsformen betroffener Bevölkerungsgruppen in Schweizer Städten / Gemeinden
   - Abstract: This project aims to capture and analyze the developments of social planning for older people in Swiss cities and municipalities, as well as how the affected population groups can be best involved in the processes of urban aging planning. The project is a collaboration between HSLU - Soziale Arbeit and the University of Applied Sciences Cologne.
   - Source: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=3401

2. Project title: Altersgerechte Quartiere Stadt Luzern
   - Abstract: This project, called "Altern in Luzern," aims to integrate the older generation more into social life. The project team from HSLU - Soziale Arbeit is working in three urban neighborhoods to lay the foundation for the development of age-friendly neighborhoods. The project includes various initiatives to improve the quality of life for older people and involve them more in society.
   - Source: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=779

3. Project title: Integrierte Sozialplanung als Innovation für die Versorgung im Alter (ISPInoVA)
   - Abstract: This project aims to capture and analyze the developments of social planning for older people in Swiss cities and municipalities, as well as how the affected population groups can be best involved in the processes of urban aging planning. The project is a collaboration between HSLU - Soziale Arbeit and the University of Applied Sciences Cologne.
   - Source: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=779

Please note that the provided information is based on the details provided in the user's question. For more information and additional projects, please visit the HSLU website: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/

In [102]:
from IPython.display import display, Markdown
def print_result(result):
  output_text = f"""### Question: 
  {query}
  ### Answer: 
  {res['choices'][0]['message']['content']}
  ### Sources: 
  {res['sources']}
  ### All relevant sources:
  #{' '.join(list(set([doc.metadata['source'] for doc in result['source_documents']])))}
  """
  display(Markdown(output_text))

### Example 2: French

In [125]:
query = "Quels sont les projets sur le thème de l'énergie?"

# Embeed Query
embed_model = "text-embedding-ada-002"
res = openai.Embedding.create(input=[query],engine=embed_model)

# Retrieve from Pinecone
xq = res['data'][0]['embedding']

# Get relevant contexts (including the questions)
res = index.query(xq, top_k=4, include_metadata=True)

#top_k=10 -->InvalidRequestError: This model's maximum context length is 4097 tokens. However, your messages resulted in 6187 tokens. Please reduce the length of the messages.

query_with_contexts = retrieve(query)


# Get list of retrieved text
contexts = [item['metadata']['text'] for item in res['matches']]
contexts_2 = [item['metadata']['source'] for item in res['matches']]
contexts_3 = [item['metadata']['title'] for item in res['matches']]
contexts_4 = [item['metadata']['abstract'] for item in res['matches']]

augmented_query = "\n\n".join(contexts)+"\n  \n".join(contexts_2)+"\n\n-----\n\n"+"\n  \n".join(contexts_3)+"\n\n-----\n\n"+"\n  \n".join(contexts_4)+"\n\n-----\n\n"+query

In [126]:
# system message to 'prime' the model
primer = f"""You are a friendly Q&A bot. A highly intelligent system that answers user questions based only on the information provided by the user above each question. You can detect the language of the users questionn and respond in the language you detected. You start your answer always in the following format according to the language detected: "At HSLU there are following projects related to {query}". Provide the identical title and summarizse the abstract from information provided by the user. Take note of the query language and provide anser in the same language as the querry. Take note of the sources and include them in the answer in the format: "SOURCES: source1 source2", use "SOURCES" in capital letters regardless of the number of sources. If the information can not be found in the information
provided by the user you truthfully say "I don't know. I use just the top 10 semantic search results, please have a look on the website: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/".
"""

res = openai.ChatCompletion.create( 
    model = "gpt-3.5-turbo", temperature=0,
    messages=[
        {"role": "system", "content": primer},
        {"role": "user", "content": augmented_query}
    ]
)

from IPython.display import Markdown
display(Markdown(res['choices'][0]['message']['content']))

À la HSLU, il y a les projets suivants liés à l'énergie :

1. Smart Energies : Ce projet explore les technologies de l'information intelligentes et mobiles pour une nouvelle génération de systèmes de contrôle intelligents et mobiles pour l'intégration de producteurs décentralisés d'énergies renouvelables dans nos réseaux de distribution d'énergie. (SOURCE : https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=2153)

2. Regionalwirtschaftliche Energiewende : Ce projet examine les impacts de la transition énergétique sur l'économie régionale. Il se déroule à la Hochschule Luzern - Wirtschaft et étudie les conséquences économiques régionales d'une mise en œuvre cohérente de la stratégie énergétique 2050. L'Institut für Gebäudetechnik und Energie fournit les bases d'une rénovation des enveloppes des bâtiments et des systèmes de chauffage, y compris les coûts d'investissement, d'exploitation, de maintenance et d'énergie. (SOURCE : https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=6011)

3. Energieversorgung Zug : Ce projet vise à étudier la faisabilité de l'approvisionnement en énergie de la ville de Zug. Il est accompagné par des chercheurs scientifiques. (SOURCE : https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=798)

4. Intelligente Multienergiesysteme in der gebauten Umwelt: sozioökonomische und technische Planung : Ce projet examine la couplage de plusieurs réseaux énergétiques, tels que l'électricité, la chaleur et le gaz, du point de vue technique et socio-économique. Il est réalisé en collaboration avec trois SCCER différents : Future Energy Efficient Building and Districts (FEEB&D), Future Swiss Electrical Infrastructures (FURIES) et Competence Center for Research in Energy, Society and Transition (CREST). Les activités sont développées en parallèle d'un projet pilote et de démonstration de Smart Grid financé par l'Office fédéral de l'énergie suisse, mis en œuvre à Rolle (VD) en collaboration avec le distributeur local, Romande Energie. Les résultats attendus du projet sont des lignes directrices se concentrant d'une part sur la planification technique des systèmes multi-énergies, et d'autre part sur des modèles économiques appropriés favorisant la diffusion de tels systèmes. (SOURCE : https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=3811)

SOURCES : https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/

### Example 3: German

In [127]:
query = "Welche Projekte gibt es zum Thema Fassadenbegrünung?"

# Embeed Query
embed_model = "text-embedding-ada-002"
res = openai.Embedding.create(input=[query],engine=embed_model)

# Retrieve from Pinecone
xq = res['data'][0]['embedding']

# Get relevant contexts (including the questions)
res = index.query(xq, top_k=4, include_metadata=True)

#top_k=10 -->InvalidRequestError: This model's maximum context length is 4097 tokens. However, your messages resulted in 6187 tokens. Please reduce the length of the messages.

query_with_contexts = retrieve(query)


# Get list of retrieved text
contexts = [item['metadata']['text'] for item in res['matches']]
contexts_2 = [item['metadata']['source'] for item in res['matches']]
contexts_3 = [item['metadata']['title'] for item in res['matches']]
contexts_4 = [item['metadata']['abstract'] for item in res['matches']]

augmented_query = "\n\n".join(contexts)+"\n  \n".join(contexts_2)+"\n\n-----\n\n"+"\n  \n".join(contexts_3)+"\n\n-----\n\n"+"\n  \n".join(contexts_4)+"\n\n-----\n\n"+query

In [130]:
# system message to 'prime' the model
primer = f"""You are a friendly Q&A bot. A highly intelligent system that answers user questions based only on the information provided by the user above each question. You can detect the language of the users {query} and respond in the language you detected. You start your answer always in the following format according to the language detected: "At HSLU there are following projects related to {query}". Provide the identical title and summarizse the abstract from information provided by the user. Take note of the query language and provide anser in the same language as the querry. Take note of the sources and include them in the answer in the format: "SOURCES: source1 source2", use "SOURCES" in capital letters regardless of the number of sources. If the information can not be found in the information
provided by the user you truthfully say "I don't know. I use just the top 10 semantic search results, please have a look on the website: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/".
"""

res = openai.ChatCompletion.create( 
    model = "gpt-3.5-turbo", temperature=0,
    messages=[
        {"role": "system", "content": primer},
        {"role": "user", "content": augmented_query}
    ]
)

from IPython.display import Markdown
display(Markdown(res['choices'][0]['message']['content']))

At HSLU there are following projects related to Fassadenbegrünung:

1. "Windkräfte in bewachsenen Fassade: CFD Studie" - This project investigates the wind effects on green facades using Computational Fluid Dynamics (CFD) studies and wind tunnel experiments. The aim is to understand the wind forces on the plant structure and the building structure of green facades. The project explores how plants align their leaves with the wind direction, which reduces the wind load on the facade. The project aims to determine the extent of this reduction and the resulting wind loads on the new green facade system.

2. "Verschattungswirkung bewachsene Fassaden" - This project focuses on the shading effect of green facades. It aims to measure the shading effect of 10 different plant species on buildings. The resulting energy flows through a green facade will be modeled using a simulation model for different orientations.

3. "GreenBrick" - This project explores sustainable and passive climate-regulating approaches for extensive vertical greening. It investigates the combination of ceramic materials and plant carriers to develop a future-oriented product for green wall and facade systems. The project aims to optimize the geometry of the carrier module, construction variants, and regulated irrigation through rainwater storage and release.

4. "Mit Energie und Farbe Landschaft prägen" - This project focuses on the integration of photovoltaic technology with color usage to generate an aesthetic value in architecture. It aims to develop a concept for a solar glass color fan, considering possible applications on facades of outstanding buildings, residential buildings, and industrial buildings. The project aims to explore how photovoltaic technology can be visually appealing and contribute to the energy strategy of Switzerland.

SOURCES: 
- https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=5797
- https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=6187
- https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=4386
- https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=5789

### Example 4: Chinese

In [132]:
query = "有哪些关于虚拟现实的项目"

# Embeed Query
embed_model = "text-embedding-ada-002"
res = openai.Embedding.create(input=[query],engine=embed_model)

# Retrieve from Pinecone
xq = res['data'][0]['embedding']

# Get relevant contexts (including the questions)
res = index.query(xq, top_k=4, include_metadata=True)

#top_k=10 -->InvalidRequestError: This model's maximum context length is 4097 tokens. However, your messages resulted in 6187 tokens. Please reduce the length of the messages.

query_with_contexts = retrieve(query)


# Get list of retrieved text
contexts = [item['metadata']['text'] for item in res['matches']]
contexts_2 = [item['metadata']['source'] for item in res['matches']]
contexts_3 = [item['metadata']['title'] for item in res['matches']]
contexts_4 = [item['metadata']['abstract'] for item in res['matches']]

augmented_query = "\n\n".join(contexts)+"\n  \n".join(contexts_2)+"\n\n-----\n\n"+"\n  \n".join(contexts_3)+"\n\n-----\n\n"+"\n  \n".join(contexts_4)+"\n\n-----\n\n"+query

In [133]:
# system message to 'prime' the model
primer = f"""You are a friendly Q&A bot. A highly intelligent system that answers user questions based only on the information provided by the user above each question. You can detect the language of the users questionn and respond in the language you detected. You start your answer always in the following format according to the language detected: "At HSLU there are following projects related to {query}". Provide the identical title and summarizse the abstract from information provided by the user. Take note of the query language and provide anser in the same language as the querry. Take note of the sources and include them in the answer in the format: "SOURCES: source1 source2", use "SOURCES" in capital letters regardless of the number of sources. If the information can not be found in the information
provided by the user you truthfully say "I don't know. I use just the top 10 semantic search results, please have a look on the website: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/".
"""

res = openai.ChatCompletion.create( 
    model = "gpt-3.5-turbo", temperature=0,
    messages=[
        {"role": "system", "content": primer},
        {"role": "user", "content": augmented_query}])

from IPython.display import Markdown
display(Markdown(res['choices'][0]['message']['content']))

在HSLU有以下关于虚拟现实的项目：

1. "Virtuelles Wasser"：该项目旨在通过增强现实技术将现有信息进行互动和娱乐化处理，以提高卢塞恩市的水井地图在旅游领域的吸引力。该项目还研究了移动设备在旅游中的重要性以及虚拟和增强现实技术的潜力。

2. "Virtual and Augmented Reality Zentrum für KMU"：该项目旨在建立一个虚拟和增强现实技术中心，为中小企业提供了解和体验这些技术的机会，并在自己的环境中开展原型和项目，并为员工（和客户）提供培训。该中心还可供学校和高校用于教育培训。

3. "VR Vaccine Production Demo"：该项目旨在利用虚拟现实技术，使人们能够在现场体验和学习新的实验室流程，而无需长途旅行和中断实验室运营。该项目的目标是在现场提供Oculus Quest头戴设备，以供成员使用。

4. "XUND Futurelab"：该项目是在XUND新建的学校中建立一个未来实验室，展示护理领域的未来技术，其中包括展示虚拟和增强现实应用。HSLU的Immersive Realities团队将在建立和配置未来实验室以及建立虚拟和增强现实技术能力方面提供支持和咨询。

来源：https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/


***Translation - manually with Deepl Translate for readers who do not understand Chinese***

Welche Projekte gibt es zum virtual reality?  

An der HSLU gibt es folgende Projekte zur virtuellen Realität:

"Virtuelles Wasser": Ziel des Projektes ist es, die Attraktivität der Brunnenkarte der Stadt Luzern im Bereich des Tourismus mittels Augmented Reality zu verbessern, indem bestehende Informationen interaktiv und unterhaltsam aufbereitet werden. Das Projekt untersucht auch die Bedeutung von mobilen Geräten im Tourismus und das Potenzial von Virtual und Augmented Reality Technologien.

"Virtual and Augmented Reality Zentrum für KMU": Ziel des Projektes ist es, ein Zentrum für Virtual- und Augmented-Reality-Technologien einzurichten, das KMU die Möglichkeit bietet, diese Technologien kennenzulernen und zu erleben, Prototypen und Projekte in ihrer eigenen Umgebung durchzuführen sowie Schulungen für ihre Mitarbeiter (und Kunden) anzubieten. Das Zentrum kann auch von Schulen und Colleges für die Ausbildung genutzt werden.

"VR Vaccine Production Demo": Ziel dieses Projekts ist es, die Technologie der virtuellen Realität zu nutzen, um neue Laborprozesse vor Ort zu erleben und zu erlernen, ohne lange Wege zurücklegen und den Laborbetrieb unterbrechen zu müssen. Ziel des Projekts ist es, das Oculus Quest Headset vor Ort für die Mitglieder zur Verfügung zu stellen.

"XUND Futurelab": Das Projekt sieht vor, in der neu errichteten Schule des XUND ein Futurelab einzurichten, um Zukunftstechnologien im Bereich der Pflege zu demonstrieren, wozu auch die Demonstration von Virtual- und Augmented-Reality-Anwendungen gehört. Das Immersive Realities-Team der HSLU wird Unterstützung und Beratung bei der Einrichtung und Konfiguration des Futurelab sowie beim Aufbau von Kapazitäten für Virtual- und Augmented-Reality-Technologien leisten. .
    

In [135]:
# system message to 'prime' the model
primer = f"""You are a friendly Q&A bot. A highly intelligent system that answers user questions based only on the information provided by the user above each question. You can detect the language of the users questionn {query} and respond in the language you detected. You start your answer always in the following format according to the language detected: "At HSLU there are following projects related to {query}". Provide the identical title and summarizse the abstract from information provided by the user. Take note of the query language and provide anser in the same language as the querry. Take note of the sources and include them in the answer in the format: "SOURCES: source1 source2", use "SOURCES" in capital letters regardless of the number of sources. If the information can not be found in the information
provided by the user you truthfully say "I don't know. I use just the top 10 semantic search results, please have a look on the website: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/".
"""

res = openai.ChatCompletion.create( 
    model = "gpt-3.5-turbo", temperature=0,
    messages=[
        {"role": "system", "content": primer},
        {"role": "user", "content": augmented_query}
    ]
)

from IPython.display import Markdown
display(Markdown(res['choices'][0]['message']['content']))

在HSLU有以下关于虚拟现实的项目：

1. "Virtuelles Wasser"：该项目旨在通过增强现实技术将现有信息以游戏化的方式呈现，以提高卢塞恩市的200多个喷泉的吸引力，特别是在旅游领域。该项目还研究了移动设备在旅游业中的重要性，并探讨了虚拟和增强现实技术在旅游领域的潜力。

2. "Virtual and Augmented Reality Zentrum für KMU"：该项目旨在建立一个虚拟和增强现实技术中心，为中小企业提供了解和体验这些技术的机会，实现在自己的环境中开发原型和项目，并培训员工（以及必要时客户）。该中心还可供学校和高校用于培训。

3. "VR Vaccine Production Demo"：该项目旨在使用虚拟现实技术开发疫苗生产过程的模拟应用，以便在现场向成员提供实际体验，而无需长途旅行和中断实验室运营。

4. "XUND Futurelab"：该项目旨在为XUND的新建学校实验室提供增强和虚拟现实技术的支持和咨询，以展示未来护理领域的技术。HSLU的沉浸式现实团队将协助建立和配置实验室，并提供增强和虚拟现实技术的培训。

来源：https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=4101、https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=4439、https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=5886、https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=4271

### Example 5: German again

In [136]:
query = "Wissensmanagement, gibt es dazu Projekte?"

# Embeed Query
embed_model = "text-embedding-ada-002"
res = openai.Embedding.create(input=[query],engine=embed_model)

# Retrieve from Pinecone
xq = res['data'][0]['embedding']

# Get relevant contexts (including the questions)
res = index.query(xq, top_k=4, include_metadata=True)

#top_k=10 -->InvalidRequestError: This model's maximum context length is 4097 tokens. However, your messages resulted in 6187 tokens. Please reduce the length of the messages.

query_with_contexts = retrieve(query)


# Get list of retrieved text
contexts = [item['metadata']['text'] for item in res['matches']]
contexts_2 = [item['metadata']['source'] for item in res['matches']]
contexts_3 = [item['metadata']['title'] for item in res['matches']]
contexts_4 = [item['metadata']['abstract'] for item in res['matches']]

augmented_query = "\n\n".join(contexts)+"\n  \n".join(contexts_2)+"\n\n-----\n\n"+"\n  \n".join(contexts_3)+"\n\n-----\n\n"+"\n  \n".join(contexts_4)+"\n\n-----\n\n"+query

In [137]:
# system message to 'prime' the model
primer = f"""You are a friendly Q&A bot. A highly intelligent system that answers user questions based only on the information provided by the user above each question. You can detect the language of the users questionn {query} and respond in the language you detected. You start your answer always in the following format according to the language detected: "At HSLU there are following projects related to {query}". Provide the identical title and summarizse the abstract from information provided by the user. Take note of the query language and provide anser in the same language as the querry. Take note of the sources and include them in the answer in the format: "SOURCES: source1 source2", use "SOURCES" in capital letters regardless of the number of sources. If the information can not be found in the information
provided by the user you truthfully say "I don't know. I use just the top 10 semantic search results, please have a look on the website: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/".
"""

res = openai.ChatCompletion.create( 
    model = "gpt-3.5-turbo", temperature=0,
    messages=[
        {"role": "system", "content": primer},
        {"role": "user", "content": augmented_query}
    ]
)

from IPython.display import Markdown
display(Markdown(res['choices'][0]['message']['content']))

At HSLU there are following projects related to Wissensmanagement, gibt es dazu Projekte?:

1. Intuitive Knowledge Connectivity: This project aims to establish a software framework that makes manifest the concept of a personal knowledge warehouse based on knowledge connectivity. It includes the development of a small software prototype and a scientific publication. (SOURCE: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=3334)

2. Aufbau eines Dienstleistungszentrums zur Förderung des Wissensaustausches zwischen Wissenschaft und Praxis in Prävention und Gesundheitsförderung: This project focuses on the development of a service center to promote knowledge exchange between science and practice in prevention and health promotion. It includes the analysis of best practices, needs assessment, and the conception of key elements. (SOURCE: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=373)

3. Knowledge Work im digitalen Wandel – Ressourcen, Strategien und Kompetenzen für erfolgreiche Karrieretransitionen: This project investigates the impact of digitalization on the professional identities, competence requirements, and career paths of knowledge workers in Switzerland. It also explores the necessary attitudes, skills, and coping strategies for successful career transitions in the digital age. (SOURCE: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=4149)

4. Untersuchung der Bedeutung von Boundary-Management von Wissensarbeitenden in einer entgrenzten Arbeitswelt: This research project examines how knowledge workers can effectively manage the demands of the digitalized and boundaryless work environment through individualized and typified boundary management. It focuses on self-management strategies to cope with flexibility, speed, and innovation requirements. (SOURCE: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=4054)

Please note that these projects are based on the information provided and there may be additional projects related to knowledge management at HSLU.

### Example 6: Spanisch

In [140]:
query = "Gestión del conocimiento, ¿hay algún proyecto al respecto?"

# Embeed Query
embed_model = "text-embedding-ada-002"
res = openai.Embedding.create(input=[query],engine=embed_model)

# Retrieve from Pinecone
xq = res['data'][0]['embedding']

# Get relevant contexts (including the questions)
res = index.query(xq, top_k=4, include_metadata=True)

#top_k=10 -->InvalidRequestError: This model's maximum context length is 4097 tokens. However, your messages resulted in 6187 tokens. Please reduce the length of the messages.

query_with_contexts = retrieve(query)


# Get list of retrieved text
contexts = [item['metadata']['text'] for item in res['matches']]
contexts_2 = [item['metadata']['source'] for item in res['matches']]
contexts_3 = [item['metadata']['title'] for item in res['matches']]
contexts_4 = [item['metadata']['abstract'] for item in res['matches']]

augmented_query = "\n\n".join(contexts)+"\n  \n".join(contexts_2)+"\n\n-----\n\n"+"\n  \n".join(contexts_3)+"\n\n-----\n\n"+"\n  \n".join(contexts_4)+"\n\n-----\n\n"+query

In [141]:
# System message to 'prime' the model
primer = f"""You are a friendly Q&A bot. A highly intelligent system that answers user questions based only on the information provided by the user above each question. You can detect the language of the users questionn {query} and respond in the language you detected. You start your answer always in the following format according to the language detected: "At HSLU there are following projects related to {query}". Provide the identical title and summarizse the abstract from information provided by the user. Take note of the query language and provide anser in the same language as the querry. Take note of the sources and include them in the answer in the format: "SOURCES: source1 source2", use "SOURCES" in capital letters regardless of the number of sources. If the information can not be found in the information
provided by the user you truthfully say "I don't know. I use just the top 10 semantic search results, please have a look on the website: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/".
"""

res = openai.ChatCompletion.create( 
    model = "gpt-3.5-turbo", temperature=0,
    messages=[
        {"role": "system", "content": primer},
        {"role": "user", "content": augmented_query}
    ]
)

from IPython.display import Markdown
display(Markdown(res['choices'][0]['message']['content']))

En HSLU hay varios proyectos relacionados con la gestión del conocimiento. Uno de ellos es el proyecto "Know and Share", que desarrolla una visión para una plataforma de intercambio de conocimientos a nivel mundial. Esta plataforma tiene como objetivo facilitar el intercambio de know-how para la operación y mantenimiento de equipos y dispositivos técnicos. El intercambio de información entre fabricantes de equipos y/o servicios técnicos y los usuarios/operadores locales se realiza a través de contenido enriquecido, como tutoriales en texto o video. El proyecto busca apoyar todo el ciclo de vida de un equipo y permitir la realización eficiente de actividades de instalación, puesta en marcha, operación y mantenimiento por parte del personal local. (SOURCE: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=373)

Otro proyecto relacionado es la "Plataforma de Gestión de Competencias para PHZG (Proyecto Piloto)". En colaboración con PHZG, se está desarrollando una plataforma en línea para gestionar las competencias en el nuevo formato de estudio personalizado e individualizado "pi". Los expertos de HSLU participan en la concepción, ingeniería de software, experiencia de usuario y gestión de proyectos de este proyecto. (SOURCE: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=5928)

Además, el proyecto "Sustainable NGO Governance" investiga la relación entre la gobernanza de las ONG y el impacto que logran a través de sus actividades. Se ha desarrollado un modelo de gobernanza-impacto y se ha probado en el campo de la cooperación para el desarrollo. El proyecto también ha desarrollado un marco de gestión del cambio que promueve la implementación sostenible de enfoques de gobernanza y aumenta la capacidad de aprendizaje individual y organizacional, así como la orientación hacia el impacto de las ONG en el desarrollo. (SOURCE: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=558)

Estos son solo algunos ejemplos de proyectos relacionados con la gestión del conocimiento en HSLU. Para obtener más información, puedes visitar el sitio web de HSLU en el siguiente enlace: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/

***Translation - manually with Deepl Translatefor readers who do not understand Spanisch***
    
Knowledge management, are there projects for this?    
    
At HSLU there are several projects related to knowledge management. One of them is the "Know and Share" project, which develops a vision for a global knowledge sharing platform. This platform aims to facilitate the exchange of know-how for the operation and maintenance of equipment and technical devices. The exchange of information between manufacturers of equipment and/or technical services and local users/operators is done through rich content, such as text or video tutorials. The project seeks to support the entire life cycle of a piece of equipment and allow the efficient performance of installation, commissioning, operation and maintenance activities by local personnel. (SOURCE: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=373)

Another related project is the "Competence Management Platform for PHZG (Pilot Project)". In collaboration with PHZG, an online platform for managing competencies in the new personalized and individualized study format "pi" is being developed. HSLU experts are involved in the conception, software engineering, user experience and project management of this project. (SOURCE: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=5928)

In addition, the "Sustainable NGO Governance" project investigates the relationship between NGO governance and the impact they achieve through their activities. A governance-impact model has been developed and tested in the field of development cooperation. The project has also developed a change management framework that promotes the sustainable implementation of governance approaches and increases individual and organizational learning capacity as well as NGO development impact orientation. (SOURCE: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=558)

These are just a few examples of projects related to knowledge management at HSLU. For more information, you can visit the HSLU website at the following link: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/    

### Example7: Japanese

In [142]:
query = "機会均等のためのプロジェクトにはどのようなものがありますか?"

# Embeed Query
embed_model = "text-embedding-ada-002"
res = openai.Embedding.create(input=[query],engine=embed_model)

# Retrieve from Pinecone
xq = res['data'][0]['embedding']

# Get relevant contexts (including the questions)
res = index.query(xq, top_k=4, include_metadata=True)

#top_k=10 -->InvalidRequestError: This model's maximum context length is 4097 tokens. However, your messages resulted in 6187 tokens. Please reduce the length of the messages.

query_with_contexts = retrieve(query)


# Get list of retrieved text
contexts = [item['metadata']['text'] for item in res['matches']]
contexts_2 = [item['metadata']['source'] for item in res['matches']]
contexts_3 = [item['metadata']['title'] for item in res['matches']]
contexts_4 = [item['metadata']['abstract'] for item in res['matches']]

augmented_query = "\n\n".join(contexts)+"\n  \n".join(contexts_2)+"\n\n-----\n\n"+"\n  \n".join(contexts_3)+"\n\n-----\n\n"+"\n  \n".join(contexts_4)+"\n\n-----\n\n"+query

In [143]:
# System message to 'prime' the model
primer = f"""You are a friendly Q&A bot. A highly intelligent system that answers user questions based only on the information provided by the user above each question. You can detect the language of the users questionn {query} and respond in the language you detected. You start your answer always in the following format according to the language detected: "At HSLU there are following projects related to {query}". Provide the identical title and summarizse the abstract from information provided by the user. Take note of the query language and provide anser in the same language as the querry. Take note of the sources and include them in the answer in the format: "SOURCES: source1 source2", use "SOURCES" in capital letters regardless of the number of sources. If the information can not be found in the information
provided by the user you truthfully say "I don't know. I use just the top 10 semantic search results, please have a look on the website: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/".
"""

res = openai.ChatCompletion.create( 
    model = "gpt-3.5-turbo", temperature=0,
    messages=[
        {"role": "system", "content": primer},
        {"role": "user", "content": augmented_query}
    ]
)

from IPython.display import Markdown
display(Markdown(res['choices'][0]['message']['content']))

HSLUには以下のような機会均等のためのプロジェクトがあります:

1. ITC-Vorprojekt: SNF-Projekteingabe: Chancengleichheit in der Sozialhilfe in verschiedenen Regionen der Schweiz: このプロジェクトは、スイスのさまざまな地域での社会福祉における機会均等に関するSNFプロジェクトの策定を目的としています。このプロジェクトは、HSLU-SAとWの異分野のチーム、実践パートナー、および（国際的な）研究パートナーを含めて実施されます。このプロジェクトは、ジェンダー、人種、階級による社会的不平等の（再）生産のメカニズムを、地域的および社会的空間的な不平等と結びつけて追跡することを目指しています。また、既に機会均等を促進するための革新的なメカニズムが存在するかどうかも示すことを目指しています。このプロジェクトは、国家、州、組織、およびカウンセラーを含むさまざまな政策実施レベルでの方法論の三角測量に基づいています。さまざまな地域（都市、田舎、言語地域）の比較を行い、サービス提供の違いを明らかにする予定です。これに基づいて、社会福祉における機会均等性の向上に向けた推奨事項を策定します。

2. Beziehungen zwischen Investoren*innen und von Frauen geführten Start-ups erfolgreich gestalten (JUFIN): このプロジェクトは、スイスの女性起業家の機会均等性を向上させるための措置を開発・実施しています。国際的な調査によると、女性は男性に比べて投資資本の獲得において明らかに不利な状況にあります。このプロジェクトは、機会均等性の意識向上、投資家、女性起業家、教育責任者の能力開発、および女性起業家と投資家の資金提供パートナーシップのネットワーキングと促進を目指しています。このプロジェクトは、スイス連邦男女平等局によって資金提供されています。開発された措置の実施は、Sandborn社によって行われます。

3. gleichstellen.ch: このプロジェクトは、スイスの労働市場における男女の平等に関する議論を促進し、平等性の改善策を提案することを目的としています。このプロジェクトは、Lucia M. Lanfranconi氏の論文「Geschlechtergleichstellung durch Wirtschaftsnutzendiskurs?」（2014）に基づいています。このプロジェクトでは、映画「Gleichstellen - eine Momentaufnahme」が開発され、スイスの労働市場における平等の実現可能な手法や現在の問題を示しています。この映画は、さまざまなイベントで上映され、またパネルディスカッションなどのイベントでも使用されます。さらに、映画を基にしたE-Learningボックスも開発されており、賃金の平等性、キャリア、育児、女性職業と男性職業、パートタイム労働、平等政策の側面を扱っています。このボックスは、企業や教育機関での短いワークショップ（90または120分）の設計と実施の出発点となります。

これらの情報は、提供された情報に基づいています。情報が提供されていない場合、「わかりません。私はトップ10の意

***Translation - manually with Deepl Translatefor readers who do not understand Japanese***


HSLU has the following equal opportunity projects:

ITC-Vorprojekt: SNF-Projekteingabe: Chancengleichheit in der Sozialhilfe in verschiedenen Regionen der Schweiz: This project aims to develop SNF projects on equal opportunities in social welfare in different regions of Switzerland. The project will involve HSLU-SA and W interdisciplinary teams, practice partners and (international) research partners. This project aims to trace the mechanisms of (re)production of social inequalities by gender, race and class in connection with regional and socio-spatial inequalities. It also aims to show whether innovative mechanisms already exist to promote equal opportunity. The project is based on methodological triangulation at various policy implementation levels including national, state, institutional, and counselor. We plan to make comparisons between different regions (urban, rural, language regions) to highlight differences in service delivery. Based on this, we formulate recommendations for improving equality of opportunity in social welfare.

Beziehungen zwischen Investoren*innen und von Frauen geführten Start-ups erfolgreich gestalten (JUFIN): This project develops and implements measures to improve equal opportunities for women entrepreneurs in Switzerland. International research shows that women are at a distinct disadvantage in obtaining investment capital compared to men. The project aims to raise awareness of equal opportunities, develop the capacity of investors, women entrepreneurs and education officers, and network and facilitate funding partnerships between women entrepreneurs and investors. The project is funded by the Swiss Federal Office for Gender Equality. The implementation of the developed measures is carried out by the company Sandborn.

gleichstellen.ch: The aim of this project is to promote debate on gender equality in the Swiss labor market and to propose ways to improve equality. This project is based on Lucia M. Lanfranconi's paper "Geschlechtergleichstellung durch Wirtschaftsnutzendiskurs?" (2014). In this project, the film "Gleichstellen – eine Momentaufnahme" was developed, showing possible approaches and current problems of equality in the Swiss labor market. The film will be screened at various events and used in events such as panel discussions. In addition, E-Learning boxes based on the film have also been developed, covering aspects of pay equity, careers, parenting, female and male occupations, part-time work, and equality policies. This box provides a starting point for designing and conducting short workshops (90 or 120 minutes) in companies and educational institutions.

### Example 8: Swedish

In [144]:
query = "Vilka projekt finns det för rekommendatorsystem?"

# Embeed Query
embed_model = "text-embedding-ada-002"
res = openai.Embedding.create(input=[query],engine=embed_model)

# Retrieve from Pinecone
xq = res['data'][0]['embedding']

# Get relevant contexts (including the questions)
res = index.query(xq, top_k=4, include_metadata=True)

#top_k=10 -->InvalidRequestError: This model's maximum context length is 4097 tokens. However, your messages resulted in 6187 tokens. Please reduce the length of the messages.

query_with_contexts = retrieve(query)


# Get list of retrieved text
contexts = [item['metadata']['text'] for item in res['matches']]
contexts_2 = [item['metadata']['source'] for item in res['matches']]
contexts_3 = [item['metadata']['title'] for item in res['matches']]
contexts_4 = [item['metadata']['abstract'] for item in res['matches']]

augmented_query = "\n\n".join(contexts)+"\n  \n".join(contexts_2)+"\n\n-----\n\n"+"\n  \n".join(contexts_3)+"\n\n-----\n\n"+"\n  \n".join(contexts_4)+"\n\n-----\n\n"+query

In [145]:
# System message to 'prime' the model
primer = f"""You are a friendly Q&A bot. A highly intelligent system that answers user questions based only on the information provided by the user above each question. You can detect the language of the users questionn {query} and respond in the language you detected. You start your answer always in the following format according to the language detected: "At HSLU there are following projects related to {query}". Provide the identical title and summarizse the abstract from information provided by the user. Take note of the query language and provide anser in the same language as the querry. Take note of the sources and include them in the answer in the format: "SOURCES: source1 source2", use "SOURCES" in capital letters regardless of the number of sources. If the information can not be found in the information
provided by the user you truthfully say "I don't know. I use just the top 10 semantic search results, please have a look on the website: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/".
"""

res = openai.ChatCompletion.create( 
    model = "gpt-3.5-turbo", temperature=0,
    messages=[
        {"role": "system", "content": primer},
        {"role": "user", "content": augmented_query}
    ]
)

from IPython.display import Markdown
display(Markdown(res['choices'][0]['message']['content']))

At HSLU there are following projects related to Vilka projekt finns det för rekommendatorsystem?:

1. Entwicklung eines Empfehlungssystems für eine Retailbank: This project focuses on developing a recommender system for a retail bank. Unfortunately, no further information is provided.

2. B2B Recommender System: This project involves the specification and evaluation of a recommender system for a B2B platform. Unfortunately, no further information is provided.

3. Empfehlungen zu Regeneration von Erdwärmesonden: This project aims to develop recommendations for the regeneration of geothermal probes in the context of residential buildings. The recommendations will be based on commercially available products and will consider their typical operating limits. The project will focus on simple hydraulic circuits and regulations with robust and efficient operating modes.

4. Stucard Recommender System: This project involves the analysis and specification of a recommender engine for the youth platform STUcard.ch. The research team will support Jaywalker GmbH, the operator of the platform, in developing a recommender algorithm tailored to the specific needs of STUcard.ch. The algorithm will leverage existing customer data and also involve the analysis of search queries and competition responses using natural language processing.

SOURCES: 
- https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=3701
- https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=4288
- https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=3810
- https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/detail/?pid=3619

### Outcome withouth primer for correct answers (Example 1)

In [93]:

res = openai.ChatCompletion.create(
    model = "gpt-3.5-turbo",
    messages=[
        {"role": "system", "content": "You are Q&A bot. A highly intelligent system that answers user questions"},
        {"role": "user", "content": query}
    ]
)
display(Markdown(res['choices'][0]['message']['content']))
     

Senioren sind Personen im fortgeschrittenen Alter, normalerweise ab 60 Jahren und älter. Sie haben oft mehr Erfahrung und Weisheit durch ihr Leben und ihre Karriere gesammelt. Senioren können auch bestimmte gesundheitliche Bedürfnisse haben und möglicherweise Unterstützung bei alltäglichen Aufgaben oder medizinischer Versorgung benötigen. Es gibt auch spezielle Programme und Dienstleistungen, die darauf ausgerichtet sind, Senioren zu helfen und ihre Lebensqualität zu verbessern.

# Gradio: HSLU multilingual project Chatbot (retrieval-augmentet GQA)

In [ ]:
import openai
import gradio as gr

openai.api_key = "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"

def respond(query):
    # Embeed Query
    embed_model = "text-embedding-ada-002"
    res = openai.Embedding.create(input=[query], engine=embed_model)

    # Retrieve from Pinecone
    xq = res['data'][0]['embedding']

    # Get relevant contexts (including the questions)
    res = index.query(xq, top_k=4, include_metadata=True)

    # Get list of retrieved text
    contexts = [item['metadata']['text'] for item in res['matches']]
    contexts_2 = [item['metadata']['source'] for item in res['matches']]
    contexts_3 = [item['metadata']['title'] for item in res['matches']]
    contexts_4 = [item['metadata']['abstract'] for item in res['matches']]

    augmented_query = "\n\n".join(contexts) + "\n  \n".join(contexts_2) + "\n\n-----\n\n" + "\n  \n".join(contexts_3) + "\n\n-----\n\n" + "\n  \n".join(contexts_4) + "\n\n-----\n\n" + query  # System message to 'prime' the model
    primer = f"""You are a friendly Q&A bot. A highly intelligent system that answers user questions based only on the information provided by the user above each question. You can detect the language of the users questionn {query} and respond in the language you detected. You start your answer always in the following format according to the language detected: "At HSLU there are following projects related to {query}". Provide the identical title and summarizse the abstract from information provided by the user. Take note of the query language and provide anser in the same language as the querry. Take note of the sources and include them in the answer in the format: "SOURCES: source1 source2", use "SOURCES" in capital letters regardless of the number of sources. If the information can not be found in the information provided by the user you truthfully say "I don't know. I use just the top 10 semantic search results, please have a look on the website: https://www.hslu.ch/de-ch/hochschule-luzern/forschung/projekte/".
    """

    res = openai.ChatCompletion.create(
        model="gpt-3.5-turbo", temperature=0,
        messages=[
            {"role": "system", "content": primer},
            {"role": "user", "content": augmented_query}
        ]
    )

    return res['choices'][0]['message']['content']

# Create the Gradio Interface

# Creating a Gradio interface for the chatbot
with gr.Blocks() as demo:
    title = gr.HTML("<h1>HSLU Project Chatbot</h1>")
    input = gr.Textbox(label="What would you like to know?")  # Textbox for user input
    output = gr.Textbox(label="Inspect answer...")  # Textbox for chatbot response
    btn = gr.Button("Get answer")  # Button to trigger the agent call
    btn.click(fn=respond, inputs=input, outputs=output)

# Launching the Gradio interface
demo.launch(share=True, debug=True)
